jolt-core/src/subprotocols/grand_product.rs

```diff
@@ -1,530 +1,1699 @@
+                //     Σ coeff[batch_index] * (Σ eq_evals[i] * left[i] * right[i])
+                // Note that if left[i] and right[i] are all 1s, the inner sum is:
+                //     Σ eq_evals[i] = eq_eval_sum
+                // To recover the actual inner sum, we find all the non-1
+                // left[i] and right[i] terms and compute the delta:
+                //     ∆ := Σ eq_evals[j] * (left[j] * right[j] - 1)    ∀j where left[j] ≠ 1 or right[j] ≠ 1
+                // Then we can compute:
+                //    coeff[batch_index] * (eq_eval_sum + ∆) = coeff[batch_index] * (Σ eq_evals[i] + Σ eq_evals[j] * (left[j] * right[j] - 1))
+                //                                           = coeff[batch_index] * (Σ eq_evals[j] * left[j] * right[j])
+                // ...which is exactly the summand we want.
+                DynamicDensityGrandProductLayer::Sparse(sparse_layer) => {
+                    // Computes:
+                    //     ∆ := Σ eq_evals[j] * (left[j] * right[j] - 1)    ∀j where left[j] ≠ 1 or right[j] ≠ 1
+                    // for the evaluation points {0, 2, 3}
+                    let mut delta = (F::zero(), F::zero(), F::zero());
+
+                    let mut next_index_to_process = 0usize;
+                    for (j, (index, value)) in sparse_layer.iter().enumerate() {
+                        if *index < next_index_to_process {
+                            // This node was already processed in a previous iteration
+                            continue;
+                        }
+                        let neighbors = [
+                            sparse_layer
+                                .get(j + 1)
+                                .cloned()
+                                .unwrap_or((index + 1, F::one())),
+                            sparse_layer
+                                .get(j + 2)
+                                .cloned()
+                                .unwrap_or((index + 2, F::one())),
+                            sparse_layer
+                                .get(j + 3)
+                                .cloned()
+                                .unwrap_or((index + 3, F::one())),
+                        ];
+
+                        let find_neighbor = |query_index: usize| {
+                            neighbors
+                                .iter()
+                                .find_map(|(neighbor_index, neighbor_value)| {
+                                    if *neighbor_index == query_index {
+                                        Some(neighbor_value)
+                                    } else {
+                                        None
+                                    }
+                                })
+                                .cloned()
+                                .unwrap_or(F::one())
+                        };
+
+                        let (left, right) = match index % 4 {
```

Comment:
Can you explain this switch? Everything else is very well documented. 

Maybe something like: "We process in chunks of 4. [left_low, left_high, right_low, right_high]. Depending on the dense index we determine which of these are present and assume the rest are sparse (1)"

Reply:
done